### ISSUE TITLE
Something like all_same?

### ISSUE THREAD

**Issue by phimuemue** [bluss/rust-itertools#175]
Do you think it would be helpful to have something that determines whether all elements of an iterator are the same?

**Comment by bluss** [comment]
Hm first approach, is there a neat way to do that already?

There is:

```rust
iter.tuple_windows().all(|(x, y)| x == y)
// but this requires cloning elements (irrelevant if they are integers or references)
```

There is (weird one):

```rust
all_equal == iter.dedup().nth(1).is_none() // only zero or one unique elements
```

IMO I haven't had the need for this operation, I can see that current itertools doesn't cover it very well. `.dedup()` is pretty close in fact, but it's not intuitive.

**Comment by phimuemue** [comment]
Thank you for your reply. Actually, I found out that in my concrete case, there is another - more natural - way avoiding `all_same`, so it is not very urgent for me.

However, I still think that it might be useful, especially within `assert!` (imagine e.g. you're using `group_by` and need a representative from each group).

I do not know how the dedup actually works, so I can not tell whether `dedup().nth(1)` solution does more work than necessary.

**Comment by bluss** [comment]
dedup is lazy and only looks at adjacent element so it should be close to doing the minimal work required (this is without me cross checking with the implementation).

**Comment by bluss** [comment]
This method needs a better name. With `.all_equal()` I think it would be a go. I'd consider implementing it using dedup.

**Comment by phimuemue** [comment]
I think you are correct that a `dedup`-based solution is convenient.

However, I saw that `dedup` internally remembers the last seen element (it works via `CoalesceCore` which has `Ok(joined) => last = joined`, i.e. assigning `last` in each internal step). I think remembering `last` might be more work than necessary, or am I mistaken?

**Comment by bluss** [comment]
I think it needs to remember last. Otherwise, why not prove it with an implementation :smile:

**Comment by phimuemue** [comment]
Should we close this?

**Comment by bluss** [comment]
Sure. Fixed in #196
